packages/@ember/routing/route.ts

```diff
@@ -1 +1,2567 @@
+    import { reject } from 'rsvp';
+    import Route from '@ember/routing/route';
+    import { action } from '@ember/object';
+
+    export default class AdminRoute extends Route {
+      beforeModel() {
+        return reject('bad things!');
+      }
+
+      @action
+      error(error, transition) {
+        // Assuming we got here due to the error in `beforeModel`,
+        // we can expect that error === "bad things!",
+        // but a promise model rejecting would also
+        // call this hook, as would any errors encountered
+        // in `afterModel`.
+
+        // The `error` hook is also provided the failed
+        // `transition`, which can be stored and later
+        // `.retry()`d if desired.
+
+        this.transitionTo('login');
+      }
+    }
+    ```
+
+    `error` actions that bubble up all the way to `ApplicationRoute`
+    will fire a default error handler that logs the error. You can
+    specify your own global default error handler by overriding the
+    `error` handler on `ApplicationRoute`:
+
+    ```app/routes/application.js
+    import Route from '@ember/routing/route';
+    import { action } from '@ember/object';
+
+    export default class ApplicationRoute extends Route {
+      @action
+      error(error, transition) {
+        this.controllerFor('banner').displayError(error.message);
+      }
+    }
+    ```
+    @event error
+    @param {Error} error
+    @param {Transition} transition
+    @since 1.0.0
+    @public
+  */
+  error?(error: Error, transition: Transition): boolean | void;
+}
+
+class Route<T = unknown>
+  extends EmberObject.extend(ActionHandler, Evented)
+  implements IRoute, Evented
+{
+  static isRouteFactory = true;
+
+  context = {} as T;
+  declare currentModel: T;
+
+  _bucketCache!: BucketCache;
+  _internalName!: string;
+
+  private _names: unknown;
+
+  _router!: EmberRouter;
+  declare _topLevelViewTemplate: any;
+  declare _environment: any;
+
+  constructor(owner: Owner) {
+    super(owner);
+
+    if (owner) {
+      let router = owner.lookup('router:main');
+      let bucketCache = owner.lookup(P`-bucket-cache:main`);
+
+      assert(
+        'ROUTER BUG: Expected route injections to be defined on the route. This is an internal bug, please open an issue on Github if you see this message!',
+        router instanceof EmberRouter && bucketCache instanceof BucketCache
+      );
+
+      this._router = router;
+      this._bucketCache = bucketCache;
+      this._topLevelViewTemplate = owner.lookup('template:-outlet');
+      this._environment = owner.lookup('-environment:main');
+    }
+  }
+
+  // Implement Evented
+  declare on: (name: string, method: ((...args: any[]) => void) | string) => this;
+  declare one: (name: string, method: string | ((...args: any[]) => void)) => this;
+  declare trigger: (name: string, ...args: any[]) => any;
+  declare off: (name: string, method: string | ((...args: any[]) => void)) => this;
+  declare has: (name: string) => boolean;
```

Comment:
For another PR, but I saw it and wanted to call it out here—since this is using `declare`, maybe we should tackle it via an `interface` merge instead?